### ISSUE TITLE
EEXIST: file already exists error thrown by file_manager.js (mkdirSync)

### ISSUE THREAD

**Issue by martin770** [angular/webdriver-manager#263]
I am using `webdriver-manager update` to download drivers for webdriver standalone and chromedriver.

To save a few seconds in my script, I used node.js to fork parallel process for each of my two updaters (webdriver standalone and chromedriver) so that they can download the files in parallel instead of being limited to a single thread. There is a race condition between the processes where if one creates the outputDir first, the other will fail with an error (EEXIST: file already exists, mkdir).

For troubleshooting, you can force the error by putting an extra `fs.mkdirSync(outputDir);` in the `makeOutputDirectory` method of the `FileManager` class in file_manager.js.

A solution would be to update the FileManager to catch the EEXIST folder and silently move on, because there's no need to create a folder that already exists. The npm package `mkdirp` includes this functionality by default.

**Comment by martin770** [comment]
This might also be a problem in `config_source.js` when it goes to find the xml files, or anywhere else in the project where mkdirSync is being used.

**Comment by cnishina** [comment]
### maybe not run two processes:

Why do you need two separate processes? My argument for not needing two processes: The download requests made for both the standalone and chromedriver produces promises and happen async. These promises are pushed to a list of promises and when the list resolves, to then write the `update-config.json` file. Since these requests are async, then we do not have to launch two processes.

Now this does not mean the code could not be improved. I think `mkdirp` would be a great fix. However, my argument above just says that for efficiency, two processes will not be that beneficial.

### mkdirp makes sense, here's a use case...

So here's a use case where I can see this being a problem:
User wants to download to CI in a set location. Let's say the location they want to use is `/tmp/selenium`. So when they call `update` they use the following command: `webdriver-manager update --out_dir /tmp/selenium`. Then they run the start command with `webdriver-manager start --out_dir /tmp/selenium` and run their Protractor tests against that. If there are several CI jobs running against a folder that was not created and they both were trying to create this directory path at the same time, it would probably run into this exact error.

I think this would definitely add value. So maybe @martin770 want to submit a PR? I'll review it.

**Comment by martin770** [comment]
Mark this as a feature request then, it is a use case for me. I'll work on a patch set. I'm using it in my development environment, and it doesn't fail for me locally because I always have the node_modules folder already created. But it does fail on Jenkins CI because they start with a fresh workspace each time.

For now I have a workaround where I create the `out_dir` and the `update-config.json dir` in each of the processes (with `mkdirp` so they don't collide) before they run `webdriver-manager update`.

I ran some timings this morning to check how much time is being saved. Here are my configurations:

    The chromedriver process passes arguments to webdriver-manager like:
        'update',
        '--out_dir=' + selenium.webdriverBinaryPath,
        '--chrome=true',
        '--versions.chrome=' + selenium.webdriverVersions.chromedriver,
        '--standalone=false',
        '--gecko=false'

    The standalone process passes arguments to webdriver-manager like:
        'update',
        '--out_dir=' + selenium.webdriverBinaryPath,
        '--standalone=true',
        '--versions.standalone=' + selenium.webdriverVersions.standalone,
        '--chrome=false',
        '--gecko=false'

    The simultaneous download passes arguments to webdriver-manager like:
        'update',
        '--out_dir=' + selenium.webdriverBinaryPath,
        '--chrome=true',
        '--versions.chrome=' + selenium.webdriverVersions.chromedriver,
        '--standalone=true',
        '--versions.standalone=' + selenium.webdriverVersions.standalone,
        '--gecko=false'

    When running in parallel, I fork two processes (one chromedriver, one standalone) and wait for them to finish.
    When running serially, I run the standalone updater, then the chromdriver updater.

Running each configuration 10 times resulted in the average time for each configuration:

    chromedriver: 2.08s (median 1.79s)
    standalone: 5.68s (median 4.61s)
    parallel: 5.13s (median 4.68s)
    serial: 5.99s (median 5.52s)
    simultaneous: 10.70 (median 10.86s)

Of course these are just rough timings, and the download times may vary depending on internet speeds and such, but it seems to indicate there is some performance gain. The thing that strikes me is the difference between running 2 processes serially vs simultaneous download (with Promise.all on one thread).

**Comment by cnishina** [comment]
Could you try this again with webdriver-manager 13? (13 does not work with Protractor 5.4.2) but I'm interested to see if this is still the issue. I think this is resolved with https://github.com/angular/webdriver-manager/blob/master/lib/cmds/update.ts#L45

When all the process are downloaded, the update command quits. Also there is no update-config.json file anymore so there shouldn't be collisions for file handles if you do the proposed parallel downloading.

**Comment by cnishina** [comment]
Also sorry for getting back to you over a year and a half later. I was moved to a different project for that time period.

**Comment by cnishina** [comment]
Actually going to close this issue since this should work now. I do swallow mkdirSync errors if the folder exists so this should be fine.

https://github.com/angular/webdriver-manager/blob/master/lib/provider/utils/cloud_storage_xml.ts#L21
https://github.com/angular/webdriver-manager/blob/master/lib/provider/utils/github_json.ts#L32
https://github.com/angular/webdriver-manager/blob/master/lib/provider/utils/file_utils.ts#L141

**Comment by cnishina** [comment]
If this is still an issue, open up a new issue and we'll discuss it there. Thanks!
